warehouse/integrations/snowflake/snowflake.go

```diff
@@ -1446,8 +1458,180 @@ func (sf *Snowflake) LoadTestTable(
+		if _, ok := rolesMap[item.granteeName]; !ok {
+			return "", false, false
+		}
+		return item.privilege, true, true
+	})
+	if hasOwnershipPrivileges := filteredPrivilegesMap["OWNERSHIP"]; hasOwnershipPrivileges {
+		return nil
+	}
+	missingPrivileges := lo.Filter(sf.config.privileges.fetchSchema.required, func(privilege string, index int) bool {
+		return !filteredPrivilegesMap[privilege]
+	})
+	if len(missingPrivileges) > 0 {
+		return fmt.Errorf("missing privileges: %v", missingPrivileges)
+	}
+	return nil
+}
+
+func (sf *Snowflake) getUserRoles(ctx context.Context) ([]string, error) {
+	user := sf.Warehouse.GetStringDestinationConfig(sf.conf, model.UserSetting)
+	sqlStatement := fmt.Sprintf("SHOW GRANTS TO USER %q;", user)
+
+	rows, err := sf.DB.QueryContext(ctx, sqlStatement)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, errors.New("no grants")
+		}
+		return nil, fmt.Errorf("fetching grants: %w", err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("getting columns: %w", err)
+	}
+
+	_, roleIndex, found := lo.FindIndexOf(columns, func(item string) bool {
+		return strings.EqualFold(item, "role")
+	})
+	if !found {
+		return nil, fmt.Errorf(`"ROLE" column not found in result: %v`, columns)
+	}
+
+	var roles []string
+	for rows.Next() {
+		rowData := make([]any, len(columns))
+		rowPointers := make([]any, len(columns))
+		for i := range rowData {
+			rowPointers[i] = &rowData[i]
+		}
+		if err := rows.Scan(rowPointers...); err != nil {
+			return nil, fmt.Errorf("scanning row: %w", err)
+		}
+		roleName, ok := rowData[roleIndex].(string)
+		if !ok {
+			return nil, fmt.Errorf("invalid data type for 'role'")
+		}
+		roles = append(roles, roleName)
+	}
+	if err := rows.Err(); err != nil {
+		return nil, fmt.Errorf("fetching grants: %w", err)
+	}
+	return roles, nil
+}
+
+func (sf *Snowflake) getShowGrantsPrivileges(ctx context.Context, sqlStatement string) ([]privilegeGrant, error) {
+	rows, err := sf.DB.QueryContext(ctx, sqlStatement)
+	if err != nil {
+		if errors.Is(err, sql.ErrNoRows) {
+			return nil, errors.New("no grants")
+		}
+		return nil, fmt.Errorf("fetching grants: %w", err)
+	}
+	defer func() { _ = rows.Close() }()
+
+	columns, err := rows.Columns()
+	if err != nil {
+		return nil, fmt.Errorf("getting columns: %w", err)
+	}
+
+	columnIndices := map[string]int{}
+	requiredColumns := [4]string{"name", "privilege", "granted_on", "grantee_name"}
+
+	for i, col := range columns {
+		lowerCol := strings.ToLower(col)
+		for _, required := range requiredColumns {
+			if lowerCol == required {
+				columnIndices[required] = i
+			}
+		}
+	}
+	for _, required := range requiredColumns {
+		if _, found := columnIndices[required]; !found {
+			return nil, fmt.Errorf(`required columns %v not found in result: %v`, requiredColumns, columns)
+		}
+	}
+
+	var privileges []privilegeGrant
+	for rows.Next() {
```

Comment:
Row reading code, is there scope of reusing common code for `getUserRoles` and `getShowGrantsPrivileges`?